rewriter/SourceRewriter.cpp

```diff
@@ -40,27 +42,53 @@ typedef std::string OpaqueStruct;
 static llvm::cl::OptionCategory
     SourceRewriterCategory("Source rewriter options");
 
+static llvm::cl::opt<std::string>
+    RootDirectory("root-directory", llvm::cl::Required,
+                  llvm::cl::cat(SourceRewriterCategory),
+                  llvm::cl::desc("<root directory for input files>"));
+
+static llvm::cl::opt<std::string>
+    OutputDirectory("output-directory", llvm::cl::Required,
+                    llvm::cl::cat(SourceRewriterCategory),
+                    llvm::cl::desc("<root directory for output files>"));
+
 static llvm::cl::opt<std::string>
     OutputPrefix("output-prefix", llvm::cl::Required,
                  llvm::cl::cat(SourceRewriterCategory),
                  llvm::cl::desc("<prefix for output files>"));
 
-static llvm::cl::opt<bool>
-    BackupSources("backup-sources", llvm::cl::Optional,
-                  llvm::cl::cat(SourceRewriterCategory),
-                  llvm::cl::desc("Back up source files before rewriting"));
-
 // Map each translation unit's filename to its pkey.
 static std::map<Filename, Pkey> file_pkeys;
 
+static std::optional<std::string> try_map_input_file(const std::string &path) {
+  bool in_root_dir = path.starts_with(RootDirectory);
```

Comment:
This requires that both paths are absolute, right? This is surprising to me, I would expect some sort of canonicalization first. Perhaps `FileManager::makeAbsolutePath()` or something with realpath of the `FileEntry`?

Reply:
I think we already required absolute paths when using `bear` but I'd have to double check.

Reply:
Unfortunately we had to canonicalize paths in this PR since using `fileinfo_iterator` was giving some relative paths for the main binary in the nginx demo. `FileManager::makeAbsolutePath` prepends the working directory of the tool but what we want is to prepend the directory of the compile command. I couldn't find an llvm function that did this so I implemented this myself when iterating over the files in the AST. Then I just saved the results of the mapping from relative paths to full paths.